FR #1124: TagDetails JSON restructure to PR #1904 (Version 2) (#2091)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/tagdetails.md

### Column Requirements

TagDetails MUST adhere to the following requirements:

* TagDetails MUST be of type JSON Object (serialized as a String where necessary).
* TagDetails MUST conform to [StringHandling](#attributes.stringhandling) requirements.
* TagDetails MUST conform to [JsonObjectFormat](#attributes.jsonobjectformat) requirements.

> * TagDetails MUST be null when [Tags](#datasets.costandusage.tags) is null, tags are not present in any other *tag sources*, and no *tag sources* are supported for any user-defined [*tag scheme*](#glossary:tag-scheme).

Reviewer: This is not a valid logical transformation of the statements. It's been a few years, but I believe this would be an affirming the consequent fallacy.

Let A be column nullability such that A indicates a column MAY be null.
Let B, C, and D be the 3 bullets in original lines 16, 17, and 18 respectively.

A → B ^ C ^ D

You've rephrase this to
B ^ C ^ D → A

A logical transformation would be
~B ∥ ~C ∥ ~D → ~A

Reviewer: ```suggestion
* TagDetails MUST NOT be null when, for any user-defined [*tag scheme*](#glossary:tag-scheme), [Tags](#datasets.costandusage.tags) is not null or tags are present in any *tag sources* which did not result in *finalized* tags or *tag sources* are supported which could have contained tags.
```
I believe this is a logical transformation, but English is more difficult to write than propositional logic. 
I had to move the `for any user-defined [*tag scheme*](#glossary:tag-scheme)` clause so that it applied to all 3 conditions. As a result, instead of doing a list with commas ( b, c, or d), I made those OR statements explicit (B or C or D).
I also slightly rephrased to make B, C, and D mutually exclusive which added clarity when consolidating into one sentence. I wasn't sure about `*tag sources* are supported which could have` vs `*tag sources* are supported which did not` because we haven't specifically defined "supported" so I felt like the former made it more clear.
Do we also need to specify that these conditions are specific to the charge record? I believe that's implied because nullability conditions are applied on a per-record basis, not on a per-column basis, but I wanted to confirm.

Author: @kk09v Heard.  I pushed this through Gemini to refine further, and here's what it suggested.  Slightly more readable.  What do you think?

```suggestion
* TagDetails MUST NOT be null when [Tags](#datasets.costandusage.tags) is not null, when tags are present in any *tag sources* which did not result in *finalized* tags, or when *tag sources* are supported for any user-defined [*tag scheme*](#glossary:tag-scheme) which could have contained tags.
```

Or, what if we just wave the white flag on making this a complicated single normative requirement and go with something more closely resembling what you'd written originally, but using our composite conventions?

```suggestion
* TagDetails MUST adhere to the following nullability requirements:
  * TagDetails MUST NOT be null when [Tags](#datasets.costandusage.tags) is not null.
  * TagDetails MUST NOT be null when tags are present in any *tag sources* which did not result in *finalized* tags.
  * TagDetails MUST NOT be null when *tag sources* are supported for any user-defined [*tag scheme*](#glossary:tag-scheme) which could have contained tags.
```

Reviewer: I think the top one is fine, but I find the bottom one easier to read.

Author: Then yeah.  Let's go with the bottom one.

* TagDetails MUST conform to the [TagDetailsObjectSchema](#schemas.datasets.costandusage.tagdetailsobjectschema) JSON Schema.

## Tag Details Object

Tag Details consists of a valid JSON object which contains objects for one or more *tag schemes*, each of which contains an object describing the tags as well as an array of eligible *tag sources* which do not contain tags.

The following section details the normative requirements for the TagDetailsObject and its nested properties. For a logical overview of the expected content, see the [Schema Structure](#datasets.costandusage.tagdetails.schemastructure) and [Object Example](#datasets.costandusage.tagdetails.objectexample) sections.

### Object Requirements

The TagDetailsObject MUST adhere to the following requirements:

* TagDetailsObject.{\*} keys MUST align to the tag prefix (up to but not including the forward slash) used to define the tag scheme in Tags.

> * TagDetailsObject.Default MUST be used to represent the unprefixed user-defined tag scheme in Tags.

Reviewer: ```suggestion
* TagDetailsObject.Default MUST be used to represent the unprefixed user-defined *tag scheme* in Tags.
```

Reviewer: Changed my mind, going to italicize tag scheme all in one go.

### Tag Scheme Object

Each tag scheme object (including `Default`) contains the following entries:

| Key | Type | Required | Description |
| :--- | :--- | :--- | :--- |
| `Tags` | Object | True | An object containing all tag keys present in the Tags column for this scheme. |
| `UntaggedSources` | Array | True | A list of sources which support this tag scheme (i.e. are taggable) that had no tags applied. |

### Tags Object (Tag Key)

The tag key object contains the properties of the *finalized tag* as well as other *tag sources* where this key was present:

| Key | Type | Required | Description |
| :--- | :--- | :--- | :--- |
| `TagSource` | String or Null | True | The type of *tag source* where the tag key is present. |
| `TagSourceId` | String or Null | True | The identifier of the *tag source* where the tag key is present. |
| `TagValue` | String, Boolean, Number, or Null | True |  The value associated with the tag key. |
| `AncestorTaggedSources` | Object or Null | True | An object containing all *tag sources* where the corresponding tag key was present which did not result in the *finalized tag*. |

> ### Ancestor Tagged Sources Object

Reviewer: ```suggestion
### Ancestor Tagged Source Object
```

This refers to each Ancestor Tagged Source object inside of the AncestorTaggedSources object.

Author: Agree; accepting.

* When TagDetailsObject.{TagScheme}.Tags.{TagKey}.AncestorTaggedSources is not null, TagDetailsObject.{TagScheme}.Tags.{TagKey}.AncestorTaggedSources adheres to the following additional requirements:
  * TagDetailsObject.{TagScheme}.Tags.{TagKey}.AncestorTaggedSources.{TagSource} key MUST denote the *tag source*.
  * TagDetailsObject.{TagScheme}.Tags.{TagKey}.AncestorTaggedSources.{TagSource}.TagSourceId MUST contain the identifier of the TagSource.
  * TagDetailsObject.{TagScheme}.Tags.{TagKey}.AncestorTaggedSources.{TagSource}.TagValue MUST adhere to the same boolean/null conditional logic as the finalized TagValue.
  * TagDetailsObject.{TagScheme}.Tags.{TagKey}.AncestorTaggedSources SHOULD contain objects for *tag sources* which did not result in the *finalized tag*.

## Schema Structure

TagDetails contains a structured JSON object defining tag eligibility and provenance from all tag sources.

### Top-Level Properties

| Property | Type | Required | Description |
| :--- | :--- | :--- | :--- |

> | `Default` | Object | True | The object containing information about the unprefixed tag scheme in the Tags column. |

Author: @kk09v What do you think about this?  I believe this is a valid change to make, but please confirm.

In the `### Top-Level Properties` table, `Default` is marked as **Required: True**. 
* **The Issue:** We explicitly fixed the JSON schema to use `"minProperties": 1` instead of strictly requiring `Default`, because a charge might *only* have provider-defined tags and no user-defined tags at all! 
* **The Fix:** Change `True` to `False` (or `Conditional`) so it doesn't contradict the JSON schema.
```suggestion
| `Default` | Object | False | The object containing information about the unprefixed tag scheme in the Tags column. |
```

Reviewer: In the default tag scheme
- it contains no tags
- because it could not have contained tags
- and neither could anything related to it.

Yep, I'm aware of this scenario, it's usually called "marketplace." Makes sense to me. 

Would this be Conditional instead of False? In the above scenario, it would not be present, otherwise it need to be present.

Author: Yeah -- conditional would be okay too (didn't have it in my suggestion, but was noted in my comment).  Let's do Conditional.

```suggestion
| `Default` | Object | Conditional | The object containing information about the unprefixed tag scheme in the Tags column. |
```

| `*{TagScheme}*` | Object | True | One or more objects containing information about a prefixed tag scheme in the Tags column. |

### Tag Scheme Object

Each tag scheme object (including `Default`) contains the following entries:

| Key | Type | Required | Description |
| :--- | :--- | :--- | :--- |
| `Tags` | Object | True | An object containing all tag keys present in the Tags column for this scheme. |

> | `UntaggedSources` | Array | True | A list of sources which support this tag scheme (i.e. are taggable) that had no tags applied. |

Author: @kk09v What do you think about this? I believe this is a valid change to make, but please confirm.

In the `### Tag Scheme Object` table, the type for `UntaggedSources` is listed simply as **Array**.
* **The Issue:** The schema and normative text explicitly state that `UntaggedSources` MUST be `null` when there are no eligible untagged sources. 
* **The Fix:** Change the type to `Array or Null` to match the convention established for `TagSource` and `TagValue`.

```suggestion
| `UntaggedSources` | Array or Null | True | A list of sources which support this tag scheme (i.e. are taggable) that had no tags applied. |
```

Reviewer: Good catch. We had debated back and forth about empty array vs null.

> ### Tags Object (Tag Key)

Author: @kk09v 
On your point about Ancestor Tagged Source needing to be singular instead of plural... Shouldn't this one then also change:

```suggestion
### Tag Key Object
```

So that:

```
### Tag Scheme Object (matches $defs.tagScheme)
### Tag Key Object (matches $defs.tagKeyObject)
### Ancestor Tagged Source Object (matches $defs.ancestorTag)
```

Reviewer: I like it.

}
```

> The corresponding Tags column would contain:

Reviewer: I personally feel that having the corresponding Tags column value in the examples helps illustrate the use to the TagDetails column and its relationship to the Tags column.

Author: Very good call.  Added Tags to the TagDetails appendix entry, and I added a link in the TagDetails column definition mentioning that we're doing side-by-side comparisons of TagDetails and Tags there.

Have a look and resolve this comment if you agree it looks good now.

Reviewer: Looks good.
